core/drivers/hisi_trng.c

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 HiSilicon Limited. */
+
+#include <initcall.h>
+#include <io.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define HTRNG_RANDATA_REG 0xF0
+#define HTRNG_BYTES 4
+
+#define POLL_PERIOD 10
+#define POLL_TIMEOUT 1000
+
+struct hisi_trng {
+	vaddr_t base;
+};
+
+static struct hisi_trng *trng_dev;
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	size_t current_len = 0;
+	uint32_t val = 0;
+
+	if (!trng_dev) {
+		EMSG("No valid TRNG device.");
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (!buf || !len) {
+		EMSG("Invalid input parameter.");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	do {
+		ret = IO_READ32_POLL_TIMEOUT(trng_dev->base + HTRNG_RANDATA_REG
+					    val, val, POLL_PERIOD, POLL_TIMEOUT)
+		if (ret) {
+			EMSG("Hardware busy.");
+			return TEE_ERROR_BUSY;
+		}
+		if (len - current_len >= HTRNG_BYTES) {
+			mencpy((uint8_t)buf + current_len, &val, HTRNG_BYTES);
+			current_len += HTRNG_BYTES;
+
+			if (current_len == len)
+				return TEE_SUCCESS;
+			continue;
+		}
+		memcpy((uint8_t)buf + current_len, &val, len - current_len);
+		current_len = len;
```

Comment:
a `break` instruction would be more explicit.
The whole seems easier to read as:
```c
	while (current_len < len) {
		ret = IO_READ32_POLL_TIMEOUT(trng_dev->base + HTRNG_RANDATA_REG
					    val, val, POLL_PERIOD, POLL_TIMEOUT)
		if (ret) {
			EMSG("Hardware busy.");
			return TEE_ERROR_BUSY;
		}

		size = MIN(HTRNG_BYTES, len - current_len);
		memcpy((uint8_t)buf + current_len, &val, size);
		current_len += size;
	}

	return TEE_SUCCESS;